### DISCUSSION TITLE
Making CTRLR for guitar devices, need help with latest addition RJM Overture

### DISCUSSION THREAD

**Discussion by verchrona** [RomanKubiak/ctrlr#643] [General]
Hi,

I have been at this Panel to control numerous Guitar devices. 4 of 5 devices work. I just got an RJM Overture pedal and did some controls for it. 

![Screenshot (182)](https://github.com/RomanKubiak/ctrlr/assets/169419087/350264d8-0294-4e5c-bdfe-bdc697d7fcb7)

The first 3 rows work perfectly with those specific units. The last row is dedicated to the RJM Overture. All the controls work perfectly. The annoying part about the pedal itself is that, everything "Drive Mode" is changed, it sends all current knob locations values. So basically, if I have the all the controls on the CTRLR Panel at noon, but the hardware is at at different settings... just changing the Drive Mode via midi will also make the Drive, Bass, Treble, Boost, Volume change values to whatever the hardware was parked at.

I figured out the solution is to: >>> everytime Drive Mode is changed via CTRLR, it reads the current values of Boost, Drive, Bass, Treble, and Volume on the CTRLR panel and also sends those values to the hardware. I wasn't able to get it working with a built-in Multi Type message. 

Is this something that needs to be done with the LUA editor? Or can I somehow make the Drive Mode modulator read and send other modulators' values when the Drive Mode is changed?

This is my panel attached.

[G-DEVICES (DRIVE MODE MUST SEND BOOST, DRIVE, BASS, TREBLE, VOLUME CC MESSAGES).zip](https://github.com/RomanKubiak/ctrlr/files/15426476/G-DEVICES.DRIVE.MODE.MUST.SEND.BOOST.DRIVE.BASS.TREBLE.VOLUME.CC.MESSAGES.zip)

**Comment by damiensellier** [comment]
On the "drive mode" knob you set the midi message to "none", it should be set to CC.

![Capture d’écran, le 2024-05-24 à 15 47 21](https://github.com/RomanKubiak/ctrlr/assets/55164385/edbf51c4-a1c6-49dc-afd7-c530d904ca8a)

All cc controls should work independently without interfering with each other especially if you assigned the different raws to different midi channels.
What you need is a way to read the current settings value and assign your ctrlr knobs at the right position. It's usually done by decoding a sysex message but I'm not sure if your pedal sends one. The pedal can also send CCs to update the value for 2way communication.

Are you sure the parameters reset on the pedal when you change the drive mode or when you start Ctrlr? in this case you need to disable either "modulator does not reacts to value change" or/and "Modulator does not output any value on startup".

![Capture d’écran, le 2024-05-24 à 15 58 51](https://github.com/RomanKubiak/ctrlr/assets/55164385/00803a65-95c1-46ca-94dd-cb34538790d7)


That would be great to know if you receive anything FROM the pedal if you change parameter values or when changing a program.
If not, It will be impossible to display the current position on the Panel and you will only be able to update the value if the Ctrlr knob is changed.

Let me know
Damien

NOTE : https://www.rjmmusic.com/download-content/Overture/Overture%20Manual-1.3.pdf

**Reply by damiensellier** [reply]
PS : from what I understand, when you change the drive mode, the pedal's other paramaters reset to different values. Do you want to keep the previously set values and lock them and sent them back to the pedal?
I guess the values when changing the mode are always reset to the same positions, so it would also be possible to display their respective values on the panel with a little LUA script. Not sure about the way or the other you want to have it working.
Parameter values Locking vs Pannel parameter values updating/following?

**Reply by verchrona** [reply]
Yes the values reset when the Drive Mode is changed. This actually occurs with the official RJM Overture Pedal Editor as well. They told me that it isn't supposed to happen, so I am emailing them with more information on the issue. I checked the midi monitor on CTRLR, it doesn't seem like the Drive Mode modulator on the panel is affecting any other values, it must be within the Overture pedal itself. As far as incoming midi messages, when I was testing, all midi to CTRLR was shut off to prevent bogus messages from feeding back.

Since the Drive mode throws off the Boost, Drive, Bass, Treble, and Volume which probably is occuring within the pedal and not from CTRLR, I would love to try this: After Drive mode is changed via CTRLR, CTRLR will also send the current CTRLR PANEL values for Boost, Drive, Bass, Treble, and Volume to the pedal in order to try to "correct" the value changes the pedal automatically does.

For example, if Boost, Drive, Bass, Treble, and Volume were set to: 42, 67, 27, 87, and 44... after Drive mode is changed to a different value, it will resend the five CC paramters at those same values 42, 67, 27, 87, and 44 once more to the unit. I am hoping this would rectify the issue. Its all I can come up with, otherwise, the unit may have some defect.

I do want the modulator to affect on startup, because of quick preset recal. I will be sending program changes to the plugin to change its entire Panel settings during different parts of a song. 

These are the video demonstrations I sent to RJM detailing the issue:
https://www.youtube.com/watch?v=jZrD1xxIEtk
https://www.youtube.com/watch?v=jqTaZ1M65WU

There are even more issues on 5-PIN midi as shown in Video 2, this may be a hardware limitation or some bug. This is quite a large issue because there is a lot of noise using USB, and I was planning to use midi to t=do all the CC changes.

**Reply by damiensellier** [reply]
no problem it's easy to push values to the desired CC with LUA when the Drive Mode is changed from the Panel.
Easy task there.

I recommand just updating the other parameters modulator to the desired value, their respective component will send the appropriate midi message accordingly, and panel GUI will follow, instead of just sending the CC value for each parameters.

**Reply by damiensellier** [reply]
I did replace your slider by a fixed slider to display the Drive Mode instead of the mod value.
and I assigned a LUA script to set the different Drive Values related to the Drive Mode. It's random value but you can put the Stock ones you see on the pedal to keep it consistant.

It works with a simple statement relative to the Drive Mode value 0-5, and push the other paramaters values accordingly:

```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
DriveModeForceValues = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

    if value == 0 then
        console("Clean Boost Mode")
        panel:getModulatorByName("OVERTURE DRIVE"):getComponent():setValue(0, true)
        panel:getModulatorByName("OVERTURE BASS"):getComponent():setValue(5, true)
        panel:getModulatorByName("OVERTURE TREBLE"):getComponent():setValue(64, true)
        panel:getModulatorByName("OVERTURE VOLUME"):getComponent():setValue(85, true)
    elseif value == 1 then
        console("Classic Mode")
        panel:getModulatorByName("OVERTURE DRIVE"):getComponent():setValue(32, true)
        panel:getModulatorByName("OVERTURE BASS"):getComponent():setValue(10, true)
        panel:getModulatorByName("OVERTURE TREBLE"):getComponent():setValue(15, true)
        panel:getModulatorByName("OVERTURE VOLUME"):getComponent():setValue(100, true)
    elseif value == 2 then
        console("Boutique Mode")
        panel:getModulatorByName("OVERTURE DRIVE"):getComponent():setValue(80, true)
        panel:getModulatorByName("OVERTURE BASS"):getComponent():setValue(90, true)
        panel:getModulatorByName("OVERTURE TREBLE"):getComponent():setValue(40, true)
        panel:getModulatorByName("OVERTURE VOLUME"):getComponent():setValue(50, true)
    elseif value == 3 then
        console("Versatile Mode")
        panel:getModulatorByName("OVERTURE DRIVE"):getComponent():setValue(60, true)
        panel:getModulatorByName("OVERTURE BASS"):getComponent():setValue(80, true)
        panel:getModulatorByName("OVERTURE TREBLE"):getComponent():setValue(0, true)
        panel:getModulatorByName("OVERTURE VOLUME"):getComponent():setValue(100, true)
    elseif value == 4 then
        console("Smooth Mode")
        panel:getModulatorByName("OVERTURE DRIVE"):getComponent():setValue(80, true)
        panel:getModulatorByName("OVERTURE BASS"):getComponent():setValue(90, true)
        panel:getModulatorByName("OVERTURE TREBLE"):getComponent():setValue(50, true)
        panel:getModulatorByName("OVERTURE VOLUME"):getComponent():setValue(120, true)
    elseif value == 5 then
        console("Crunch Mode")
        panel:getModulatorByName("OVERTURE DRIVE"):getComponent():setValue(120, true)
        panel:getModulatorByName("OVERTURE BASS"):getComponent():setValue(12, true)
        panel:getModulatorByName("OVERTURE TREBLE"):getComponent():setValue(100, true)
        panel:getModulatorByName("OVERTURE VOLUME"):getComponent():setValue(80, true)
    end -- filter value

end
```

Everytime you change the Drive Mode it sends the preset values as CC messages:
```
14:58:51.290	To MIDI output port 0	Control	5	Controller 104	42
14:58:51.290	To MIDI output port 0	Control	5	Effect Control 1 (coarse)	80
14:58:51.290	To MIDI output port 0	Control	5	Effect Control 2 (coarse)	90
14:58:51.290	To MIDI output port 0	Control	5	General Purpose 2 (coarse)	40
14:58:51.290	To MIDI output port 0	Control	5	Channel Volume (coarse)	50
```

[G-DEVICES_2_1_G-DEVICES_2024-05-25_14-54.panel.zip](https://github.com/RomanKubiak/ctrlr/files/15444653/G-DEVICES_2_1_G-DEVICES_2024-05-25_14-54.panel.zip)

**Reply by damiensellier** [reply]
And The other way around, if you want to keep the current Overture Panel values and override the pedal new values:

You need to store the current panel values then send them back as CC to the pedal. 

```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
DriveModeKeepsValues = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

-- Storing the current Overture Drive Values
local overtureDriveValue = panel:getModulatorByName("OVERTURE DRIVE"):getComponent():getValue()
console("overtureDriveValue : "..overtureDriveValue)
local overtureBassValue = panel:getModulatorByName("OVERTURE BASS"):getComponent():getValue()
console("overtureBassValue : "..overtureBassValue)
local overtureTrebleValue = panel:getModulatorByName("OVERTURE TREBLE"):getComponent():getValue()
console("overtureTrebleValue : "..overtureTrebleValue)
local overtureVolumeValue = panel:getModulatorByName("OVERTURE VOLUME"):getComponent():getValue()
console("overtureVolumeValue : "..overtureVolumeValue)

    if value == 0 then
        console("Clean Boost Mode")
    elseif value == 1 then
        console("Classic Mode")
    elseif value == 2 then
        console("Boutique Mode")
    elseif value == 3 then
        console("Versatile Mode")
    elseif value == 4 then
        console("Smooth Mode")
    elseif value == 5 then
        console("Crunch Mode")
    end -- filter value

    panel:getModulatorByName("OVERTURE DRIVE"):setModulatorValue(overtureDriveValue, true, true, true)
    panel:getModulatorByName("OVERTURE BASS"):setModulatorValue(overtureBassValue, true, true, true)
    panel:getModulatorByName("OVERTURE TREBLE"):setModulatorValue(overtureTrebleValue, true, true, true)
    panel:getModulatorByName("OVERTURE VOLUME"):setModulatorValue(overtureVolumeValue, true, true, true)

end
```

As you can see, the panel send the displayed values back to the pedal, so the buggy new values of the pedal while the Mode changes are going to be overriden by the old one stored on the panel. Transparent.

```
15:18:51.407	To MIDI output port 0	Control	5	Controller 104	21
15:18:51.407	To MIDI output port 0	Control	5	Effect Control 1 (coarse)	72
15:18:51.407	To MIDI output port 0	Control	5	Effect Control 2 (coarse)	46
15:18:51.407	To MIDI output port 0	Control	5	General Purpose 2 (coarse)	2
15:18:51.407	To MIDI output port 0	Control	5	Channel Volume (coarse)	127
```

[G-DEVICES_2_3_G-DEVICES_2024-05-25_15-39.panel.zip](https://github.com/RomanKubiak/ctrlr/files/15444795/G-DEVICES_2_3_G-DEVICES_2024-05-25_15-39.panel.zip)

**Reply by damiensellier** [reply]
EDIT: I did a minor change with setModulatorValue(overtureDriveValue, true, true, true) instead of setValue(overtureDriveValue, true)...

It's working everytime now.

[G-DEVICES_2_3_G-DEVICES_2024-05-25_15-39.panel.zip](https://github.com/RomanKubiak/ctrlr/files/15444795/G-DEVICES_2_3_G-DEVICES_2024-05-25_15-39.panel.zip)

**Reply by damiensellier** [reply]
NOTE: if you don't want to have the overture drive bass treble and volume sent twice on the panel startup, you have to filter the source in the LUA script ( if source == 4 then ) :
```

--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
DriveModeKeepsValues = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

if source == 4 then -- should stop this code running on startup, will work only when user interacts with GUI

-- Storing the current Overture Drive Values
local overtureDriveValue = panel:getModulatorByName("OVERTURE DRIVE"):getComponent():getValue()
console("overtureDriveValue : "..overtureDriveValue)
local overtureBassValue = panel:getModulatorByName("OVERTURE BASS"):getComponent():getValue()
console("overtureBassValue : "..overtureBassValue)
local overtureTrebleValue = panel:getModulatorByName("OVERTURE TREBLE"):getComponent():getValue()
console("overtureTrebleValue : "..overtureTrebleValue)
local overtureVolumeValue = panel:getModulatorByName("OVERTURE VOLUME"):getComponent():getValue()
console("overtureVolumeValue : "..overtureVolumeValue)

    if value == 0 then
        console("Clean Boost Mode")
    elseif value == 1 then
        console("Classic Mode")
    elseif value == 2 then
        console("Boutique Mode")
    elseif value == 3 then
        console("Versatile Mode")
    elseif value == 4 then
        console("Smooth Mode")
    elseif value == 5 then
        console("Crunch Mode")
    end -- filter value

    panel:getModulatorByName("OVERTURE DRIVE"):setModulatorValue(overtureDriveValue, true, true, true)
    panel:getModulatorByName("OVERTURE BASS"):setModulatorValue(overtureBassValue, true, true, true)
    panel:getModulatorByName("OVERTURE TREBLE"):setModulatorValue(overtureTrebleValue, true, true, true)
    panel:getModulatorByName("OVERTURE VOLUME"):setModulatorValue(overtureVolumeValue, true, true, true)

end -- filter source

end
```

**Reply by verchrona** [reply]
This is absolutely incredible. I didn't expect the code to have to be that long. I greatly appreciate you taking the time to write this code to implement with the hardware unit. It actually fixes the bug that I believe is built into the Overture pedal. 

The pedal itself is rather finicky and only connects via USB whenever I'm lucky haha. What's strange about the Overture's USB driver is that once a DAW occupies its Midi ports, and even if you shut down the DAW after that, the midi ports will remain occupied until the computer is restarted. I tried this with a couple programs and it doesn't free itself after the program using it is closed.

I realized the BOOST knob actually requires the same implementation as the other 4 knobs. I went ahead and did what you did to the other knobs and applied it to the BOOST knob as well. I modified the DriveModeKeepsValues lua page only, and didn't update the DriveModeForceValues, because I was unsure what to put after "setvalue(number, true)". But I think its working without changing the DriveModeForceValues. Is it ok not to update that Lua page or does it have to be done to have the BOOST modulator behave like the Drive, Bass, Treble, and Volume?

The updated panel is attached:
[G-DEVICES (BOOST REQUIRES SAME IMPLEMENTATION).zip](https://github.com/RomanKubiak/ctrlr/files/15446078/G-DEVICES.BOOST.REQUIRES.SAME.IMPLEMENTATION.zip)

**Reply by damiensellier** [reply]
You're welcome!
You have 2 different scripts on the panel I posted covering both options. You should be able to work from there now that you have the different required LUA functions.
Don't hesitate to ask if you have other questions during your project.

If you prefer using a midi din 5 pin instead of USB, you can get a cheap USB MIDI cable since there's no SYSEX messages. They work fine with short compliant MIDI messages. Maybe you already have some MIDI ports on your inteface ou should try if the pedal works better with it.

Enjoy!

Damien

![Capture d’écran, le 2024-05-26 à 00 37 43](https://github.com/RomanKubiak/ctrlr/assets/55164385/cd4b6dee-215d-43b7-8483-1e46fd8b6415)
![Capture d’écran, le 2024-05-26 à 00 37 54](https://github.com/RomanKubiak/ctrlr/assets/55164385/2edb9804-f393-47b8-b4f3-d059ef5c512d)

**Reply by damiensellier** [reply]
@verchrona 

`I modified the DriveModeKeepsValues lua page only, and didn't update the DriveModeForceValues, because I was unsure what to put after "setvalue(number, true)". But I think its working without changing the DriveModeForceValues. Is it ok not to update that Lua page or does it have to be done to have the BOOST modulator behave like the Drive, Bass, Treble, and Volume?`

You can even remove the one you don't need. 
The script is assigned when you clicked on the "Drive Mod" knob with the property pane open and go to "called when the modulator value changes" : 
![Capture d’écran, le 2024-05-26 à 00 46 22](https://github.com/RomanKubiak/ctrlr/assets/55164385/5a2e9cf0-c45a-48c8-ad17-961ad81d2549)

If you want to remove a LUA Method just open the LUA method editor, right-click the method to remove and well... remove it :)
![Capture d’écran, le 2024-05-26 à 00 49 50](https://github.com/RomanKubiak/ctrlr/assets/55164385/311b5bd0-ba1d-4000-953e-afef5aa7b0a5)

**Reply by damiensellier** [reply]
NOTE: you can also have values you like as default for the different Drive Modes sent to the pedal with the other method. I wrote some statements **if ... then** for the different Modes so, depending on the Mode you switch, you can have different values set up.

**Reply by verchrona** [reply]
Ah, now I understand how choosing the script for the modulator works with the drop down menu. Thanks again for everything!
